### PR TITLE
feat(restore): retry updating restore status when restore job succeeded

### LINKED ISSUE #4334: Add retry logic on updating restore
## Feature Request

**Is your feature request related to a problem? Please describe:**

logs show that restore job has complete successfully but failed to update restore cr' status due to timeout connecting to kube-apiserver. 
```
2021-12-28 20:07:46 | Error from server: context deadline exceeded
-- | --
  |   | 2021-12-28 20:07:46 | E1228 12:07:46.143592       9 restore_status_updater.go:79] Failed to update resotre [tidbxxx/xxx], error: context deadline exceeded
  |   | 2021-12-28 20:07:12 | I1228 12:07:12.139149       9 manager.go:289] restore cluster tidbxxx/xxx from full succeed
  |   | 2021-12-28 20:07:12 | I1228 12:07:12.139129       9 manager.go:274] reset cluster tidbxxx/xxx tikv_gc_life_time to 10m0s success
  |   | 2021-12-28 20:07:12 | I1228 12:07:12.132340       9 restore.go:105] Restore data for cluster tidbxxx/xxx successfully
```

**Describe the feature you'd like:**

Refer to https://github.com/pingcap/tidb-operator/blob/383d6f9d283733bf0ec403324b361c5a00d2b3f7/cmd/backup-manager/app/restore/manager.go#L298
we can make it to retry on error to gurantee this will be updated.

### PR DESCRIPTION
**Pull request by just1900**
Signed-off-by: just1900 <legendj228@gmail.com>

### What problem does this PR solve?

fix #4334 
### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [x] Manual test 
FYI, I've test it with creating a backup and then add firewall that simply drop the packets to the apiserver after br succeed, here are the retry logs( focus on "Failed to update backup ..."). After retrying 3 times, I remove the firewall and it succeeds.

  ```
  I1231 05:29:43.876251       8 manager.go:321] Get cluster tidb-operator-test/basic2-backup-s3 commitTs 430159582415028225 success
  E1231 05:30:27.684019       8 backup_status_updater.go:89] Failed to update backup [tidb-operator-test/basic2-backup-s3], error: Put "https://10.0.0.1:443/apis/pingcap.com/v1alpha1/namespaces/tidb-operator-test/backups/basic2-backup-s3": http2: client connection lost
  W1231 05:30:27.684049       8 reflector.go:424] k8s.io/client-go@v0.19.16/tools/cache/reflector.go:156: watch of *v1alpha1.Backup ended with: an error on the server ("unable to decode an event from the watch stream: http2: client connection lost") has prevented the request from succeeding
  E1231 05:30:57.696650       8 backup_status_updater.go:89] Failed to update backup [tidb-operator-test/basic2-backup-s3], error: Put "https://10.0.0.1:443/apis/pingcap.com/v1alpha1/namespaces/tidb-operator-test/backups/basic2-backup-s3": dial tcp 10.0.0.1:443: i/o timeout
  I1231 05:30:58.836848       8 trace.go:205] Trace[1474941318]: "Reflector ListAndWatch" name:k8s.io/client-go@v0.19.16/tools/cache/reflector.go:156 (31-Dec-2021 05:30:28.834) (total time: 30001ms):
  Trace[1474941318]: [30.001983021s] [30.001983021s] END
  E1231 05:30:58.836960       8 reflector.go:127] k8s.io/client-go@v0.19.16/tools/cache/reflector.go:156: Failed to watch *v1alpha1.Backup: failed to list *v1alpha1.Backup: Get "https://10.0.0.1:443/apis/pingcap.com/v1alpha1/namespaces/tidb-operator-test/backups?resourceVersion=79880584": dial tcp 10.0.0.1:443: i/o timeout
  I1231 05:31:12.751950       8 trace.go:205] Trace[336122540]: "Reflector ListAndWatch" name:k8s.io/client-go@v0.19.16/tools/cache/reflector.go:156 (31-Dec-2021 05:31:00.542) (total time: 12209ms):
  Trace[336122540]: ---"Objects listed" 12209ms (05:31:00.751)
  Trace[336122540]: [12.209709474s] [12.209709474s] END
  I1231 05:31:12.756815       8 backup_status_updater.go:86] Backup: [tidb-operator-test/basic2-backup-s3] updated successfully
  ```
  Check status of `Backup` by kubeclt get backup basic2-backup-s3 :

  ```
  NAME               STATUS     BACKUPPATH                      BACKUPSIZE   COMMITTS             AGE
  basic2-backup-s3   Complete   s3://basic-tidb-bucket/basic2   161 kB       430159582415028225   3m25s
  ```

  And here are the 5 times retry logs for restore job that I fail apiserver directly.
  
  ```
  restore-basic-restore-s3-2djks restore I1231 05:36:47.999681       8 restore.go:105] Restore data for cluster tidb-operator-test/basic-restore-s3 successfully
  restore-basic-restore-s3-2djks restore I1231 05:36:47.999734       8 manager.go:289] restore cluster tidb-operator-test/basic-restore-s3 from  succeed
  restore-basic-restore-s3-2djks restore E1231 05:37:18.001843       8 restore_status_updater.go:81] Failed to update resotre [tidb-operator-test/basic-restore-s3], error: Put "https://10.0.0.1:443/apis/pingcap.com/v1alpha1/namespaces/tidb-operator-test/restores/basic-restore-s3": dial tcp 10.0.0.1:443: connect: connection refused
  restore-basic-restore-s3-2djks restore E1231 05:37:18.013810       8 restore_status_updater.go:81] Failed to update resotre [tidb-operator-test/basic-restore-s3], error: Put "https://10.0.0.1:443/apis/pingcap.com/v1alpha1/namespaces/tidb-operator-test/restores/basic-restore-s3": dial tcp 10.0.0.1:443: connect: connection refused
  restore-basic-restore-s3-2djks restore E1231 05:37:18.025884       8 restore_status_updater.go:81] Failed to update resotre [tidb-operator-test/basic-restore-s3], error: Put "https://10.0.0.1:443/apis/pingcap.com/v1alpha1/namespaces/tidb-operator-test/restores/basic-restore-s3": dial tcp 10.0.0.1:443: connect: connection refused
  restore-basic-restore-s3-2djks restore E1231 05:37:18.037810       8 restore_status_updater.go:81] Failed to update resotre [tidb-operator-test/basic-restore-s3], error: Put "https://10.0.0.1:443/apis/pingcap.com/v1alpha1/namespaces/tidb-operator-test/restores/basic-restore-s3": dial tcp 10.0.0.1:443: connect: connection refused
  restore-basic-restore-s3-2djks restore E1231 05:37:18.049867       8 restore_status_updater.go:81] Failed to update resotre [tidb-operator-test/basic-restore-s3], error: Put "https://10.0.0.1:443/apis/pingcap.com/v1alpha1/namespaces/tidb-operator-test/restores/basic-restore-s3": dial tcp 10.0.0.1:443: connect: connection refused
  restore-basic-restore-s3-2djks restore The connection to the server 10.0.0.1:443 was refused - did you specify the right host or port?
  ```

  Check status of `Restore` :
  
  ```
  NAME               STATUS    COMMITTS   AGE
  basic-restore-s3   Running              77s
  ```

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Add retry for the status update for the Backup and Restore CR
```

### CODE REVIEW

File: cmd/backup-manager/app/restore/manager.go [hunk 1]
```diff
@@ -295,8 +296,11 @@ func (rm *Manager) performRestore(ctx context.Context, restore *v1alpha1.Restore
 		TimeCompleted: &metav1.Time{Time: finish},
 		CommitTs:      &ts,
 	}
-	return rm.StatusUpdater.Update(restore, &v1alpha1.RestoreCondition{
-		Type:   v1alpha1.RestoreComplete,
-		Status: corev1.ConditionTrue,
-	}, updateStatus)
+	// try best effort to guarantee restore status is updated.
+	return retry.OnError(retry.DefaultRetry, func(e error) bool { return e != nil }, func() error {
```
Comment (DanielZhangQD) [inline cmd/backup-manager/app/restore/manager.go:300]: Is it better to define a function `rm.StatusUpdater.RetryUpdate()`?

Comment (DanielZhangQD) [reply cmd/backup-manager/app/restore/manager.go:300]: BTW, there are many codes that will update the status for the backup and restore CR (`grep -r "Update(" cmd`), maybe it's better to update the logic of the `Update` function.

Comment (just1900) [reply cmd/backup-manager/app/restore/manager.go:300]: I'm trying to update the logic inside `Update` function at first, but I think it's not the client's responsibilty to retry on errors like conn timeout.

Do we need to make `Update` function to have the `guaranteeUpdate` semantics here?

**Comment by DanielZhangQD** [comment]
@just1900 Please verify the code change and update the test result in the PR description.

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Review by csuzhangxc — APPROVED** [review]

**Review by KanShiori — APPROVED** [review]

**Pull request merged**